### PR TITLE
CLU C# Sdk Preview

### PR DESCRIPTION
**Pull request by ZulaMostafa**
This PR contains the following for 2022-03-01 clu preview: 
- Swagger regeneration
- Tests and Samples
- Readme modification

### CODE REVIEW

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/samples/Sample4_AnalyzeConversationWithOptions.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
+# Analyze a conversation
+
+This sample demonstrates how to analyze an utterance. To get started, you'll need to create a Cognitive Language service endpoint and an API key. See the [README](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/cognitivelanguage/Azure.AI.Language.Conversations/README.md) for links and instructions.
```
Comment (heaths) [inline sdk/cognitivelanguage/Azure.AI.Language.Conversations/samples/Sample4_AnalyzeConversationWithOptions.md:3]: Perhaps, to distinguish from other samples:
```suggestion
This sample demonstrates how to analyze an utterance with additional options. To get started, you'll need to create a Cognitive Language service endpoint and an API key. See the [README](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/cognitivelanguage/Azure.AI.Language.Conversations/README.md) for links and instructions.
```

File: sdk/cognitivelanguage/Azure.AI.Language.sln [hunk 2]
```diff
@@ -22,6 +19,10 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Azure.AI.Language.Conversat
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Azure.AI.Language.Conversations.Tests", "Azure.AI.Language.Conversations\tests\Azure.AI.Language.Conversations.Tests.csproj", "{E30B7233-76F8-4966-9CDC-86D9C5B37462}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Azure.Core", "..\core\Azure.Core\src\Azure.Core.csproj", "{7DBAEA65-0607-404C-A326-8089FC2887CF}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker", "..\cognitiveservices\Knowledge.QnAMaker\src\Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker.csproj", "{46AFA6FD-6AC0-4B23-8620-7340120BAA18}"
+EndProject
```
Comment (heaths) [inline sdk/cognitivelanguage/Azure.AI.Language.sln:25]: Why was this added? Unless it's used for migration samples (and I don't see any), we typically don't want track 1 libraries in our solutions (waste of time to build).

Also, don't add Azure.Core to the solution unless required. I know there are some cases where it is, and if it was causing problems not being in the solution then, by all means, add it. :smile:

**Review by heaths — CHANGES_REQUESTED** [review]
The method signature should probably remain the same. The rest of the comments are mostly nits.

File: sdk/cognitivelanguage/Azure.AI.Language.sln (hunk 2 above)
Comment (ZulaMostafa) [reply sdk/cognitivelanguage/Azure.AI.Language.sln:25]: I was getting errors because of not adding it, will remove it once we make sure that there are no changes.

Comment (heaths) [reply sdk/cognitivelanguage/Azure.AI.Language.sln:25]: If Azure.Core was necessary, go ahead and leave it in. Or were you talking about the track 1 library?

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/AnalyzeConversationOptions.cs [hunk 3]
```diff
@@ -7,18 +7,10 @@
 namespace Azure.AI.Language.Conversations
 {
     /// <summary> The request body. </summary>
-    [CodeGenModel("ConversationAnalysisOptions")]
-    [CodeGenSuppress("AnalyzeConversationOptions", typeof(string))]
-    public partial class AnalyzeConversationOptions
+    public partial class ConversationAnalysisOptions
```
Comment (heaths) [inline sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/AnalyzeConversationOptions.cs:10]: The class needs to follow the method name + "Options", so this needs to remain `AnalyzeConversationOptions`.

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/ConversationAnalysisOptions.cs [hunk 4]
```diff
@@ -22,7 +22,11 @@ public enum ServiceVersion
             /// <summary>
             /// Service version "2021-11-01-preview".
             /// </summary>
-            V2022_03_01_Preview = 1,
+            V2021_11_01_Preview = 1,
+            /// <summary>
+            /// Service version "2022-03-01-preview".
+            /// </summary>
+            V2022_03_01_Preview = 2
```
Comment (heaths) [inline sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/ConversationAnalysisOptions.cs:29]: 1. Use trailing commas for easier diffing in the future.
2. Our guidelines are that client options are always like `ConversationAnalysisClientOptions`. See https://azure.github.io/azure-sdk/general_design.html.

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/ConversationAnalysisClient.cs [hunk 5]
```diff
@@ -74,25 +74,26 @@ protected ConversationAnalysisClient()
...
         /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> to cancel the request.</param>
-        /// <exception cref="ArgumentNullException"><paramref name="conversationItemBase"/> is null.</exception>
+        /// <exception cref="ArgumentNullException"><paramref name="input"/> is null.</exception>
         /// <exception cref="ArgumentNullException"><paramref name="project"/> is null.</exception>
         /// <exception cref="RequestFailedException">The service returned an error. The exception contains details of the service error.</exception>
-        public virtual async Task<Response<AnalyzeConversationTaskResult>> AnalyzeConversationAsync(ConversationItemBase conversationItemBase, ConversationsProject project, AnalyzeConversationOptions analysisInput = null, bool? verbose = null, CancellationToken cancellationToken = default)
+        public virtual async Task<Response<AnalyzeConversationTaskResult>> AnalyzeConversationAsync(ConversationItemBase input, ConversationsProject project, ConversationAnalysisOptions options = null, CancellationToken cancellationToken = default)
```
Comment (heaths) [inline sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/ConversationAnalysisClient.cs:84]: Why the change from a single string utterance and pack everything not required into the optional `options`?

https://docs.microsoft.com/dotnet/api/azure.ai.language.conversations.conversationanalysisclient.analyzeconversation

This is quite the departure from the original, simple design. For .NET, all *required* options are required parameters and we pack everything optional into a `{MethodName}Options` class. Having this base class doesn't make it easy to use. How are customers supposed to know what derivative classes to even pass? Can we just keep this as the simple `string utterance` like the old signature?

/cc @annatisch

**Review by heaths — CHANGES_REQUESTED** [review]
I don't think `AnalyzeConversation` should be changed. Was there a driving reason for this?

Also, the method and client options classes shouldn't have been renamed. They follow our guidelines (see link in comments).

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/ConversationAnalysisClient.cs (hunk 5 above)
Comment (ZulaMostafa) [reply sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/ConversationAnalysisClient.cs:84]: `participantId`, `id`, and `modality` which are in `ConversationItemBase` are all required, as I understand from the swagger: 

```json
"ConversationItemBase":{
   "type":"object",
   "discriminator":"modality",
   "description":"The abstract base for a user input formatted conversation (e.g., Text, Transcript).",
   "properties":{
      "participantId":{
         "description":"The participant Id of a ConversationItem",
         "type":"string"
      },
      "id":{
         "description":"The Id of a ConversationItem",
         "type":"string"
      },
      "language":{
         "description":"The input language of a ConversationItem in BCP 47 language representation.",
         "type":"string"
      },
      "modality":{
         "description":"The modality (format) of ConversationItem (e.g., Text, Transcript)",
         "type":"string",
         "enum":[
            "text"
         ],
         "x-ms-enum":{
            "name":"modality",
            "modelAsString":true
         }
      }
   },
   "required":[
      "participantId",
      "id",
      "modality"
   ]
}
```

I wanted to only keep `utterance` in the parameters but couldn't because of the required parameters, please correct me if  I am wrong.

Comment (ZulaMostafa) [reply sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/ConversationAnalysisClient.cs:84]: There was a thread about those parameters as they should not be required parameters, I will add default values for them and revert back to utterance.

File: sdk/cognitivelanguage/Azure.AI.Language.sln (hunk 2 above)
Comment (ZulaMostafa) [reply sdk/cognitivelanguage/Azure.AI.Language.sln]: I was experimenting for both of them because I had errors. But now I tried again and only Azure.Core is necessary to eliminate the errors.

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/README.md [hunk 6]
```diff
@@ -113,18 +108,15 @@ foreach (ConversationEntity entity in conversationPrediction.Entities)
...
         {
-            if (resolution is DateTimeResolution)
-            {
-                DateTimeResolution dateTimeResolution = resolution as DateTimeResolution;
-                Console.WriteLine($"Datetime Sub Kind: {dateTimeResolution.DateTimeSubKind}");
-                Console.WriteLine($"Timex: {dateTimeResolution.Timex}");
-                Console.WriteLine($"Value: {dateTimeResolution.Value}");
-                Console.WriteLine();
-            }
+            DateTimeResolution dateTimeResolution = resolution as DateTimeResolution;
```
Comment (heaths) [inline sdk/cognitivelanguage/Azure.AI.Language.Conversations/README.md:115]: Nit: if `resolution` is not a `DateTimeResolution` it will be `null` and all the accesses below will throw a `NullReferenceException`. Instead, I'd use a cast here which will at least throw a more obvious `InvalidCastException`:

```csharp
DateTimeResolution dateTimeResolution = (DateTimeResolution)resolution;
```

Comment (heaths) [reply sdk/cognitivelanguage/Azure.AI.Language.Conversations/README.md:115]: Actually, I missed in context line 113. You can do this all more simply and succinctly:

```csharp
if (resolution is DateTimeResolution dateTimeResolution)
{
  Console.WriteLine($"Datetime Sub Kind: {dateTimeResolution.DateTimeSubKind}");
  ...
}
```

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/README.md [hunk 7]
```diff
@@ -200,18 +188,20 @@ The language property in the `TextConversationItem` can be used to specify the l
 
 ```C# Snippet:ConversationAnalysis_AnalyzeConversationWithLanguage
 TextConversationItem input = new TextConversationItem(
-    participantId: "1",
-    id: "1",
-    text: "Tendremos 2 platos de nigiri de salmón braseado.")
+     participantId: "1",
+     id: "1",
+     text: "Tendremos 2 platos de nigiri de salmón braseado.")
```
Comment (heaths) [inline sdk/cognitivelanguage/Azure.AI.Language.Conversations/README.md:193]: Nit: ideally just 4 spaces, as is normal in our repo. This has 5.

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/samples/Sample3_AnalyzeConversationWithLanguage.md [hunk 8]
```diff
@@ -17,18 +17,20 @@ Once you have created a client, you can call synchronous or asynchronous methods
 
 ```C# Snippet:ConversationAnalysis_AnalyzeConversationWithLanguage
 TextConversationItem input = new TextConversationItem(
-    participantId: "1",
-    id: "1",
-    text: "Tendremos 2 platos de nigiri de salmón braseado.")
+     participantId: "1",
+     id: "1",
+     text: "Tendremos 2 platos de nigiri de salmón braseado.")
```
Comment (heaths) [inline sdk/cognitivelanguage/Azure.AI.Language.Conversations/samples/Sample3_AnalyzeConversationWithLanguage.md:22]: Same here with 5 spaces. Should just be 4.

Comment (heaths) [reply sdk/cognitivelanguage/Azure.AI.Language.Conversations/samples/Sample3_AnalyzeConversationWithLanguage.md:22]: A few other places below.

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/samples/Sample4_AnalyzeConversationWithOptions.md [hunk 9]
```diff
@@ -31,8 +31,7 @@ ConversationsProject conversationsProject = new ConversationsProject("Menu", "pr
 Response<AnalyzeConversationTaskResult> response = client.AnalyzeConversation(
     textConversationItem,
```
Comment (heaths) [inline sdk/cognitivelanguage/Azure.AI.Language.Conversations/samples/Sample4_AnalyzeConversationWithOptions.md:32]: Shouldn't this just be a string now?

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/ConversationAnalysisOptions.cs [hunk 10]
```diff
@@ -26,7 +26,7 @@ public enum ServiceVersion
             /// <summary>
             /// Service version "2022-03-01-preview".
             /// </summary>
-            V2022_03_01_Preview = 2
+            V2022_03_01_Preview = 2,
```
Comment (heaths) [inline sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/ConversationAnalysisOptions.cs:29]: This file should be called `ConversationAnalysisClientOptions.cs` to match the class name.

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/ConversationAnalysisClient.cs [hunk 11]
```diff
@@ -74,21 +74,22 @@ protected ConversationAnalysisClient()
...
-        /// <exception cref="ArgumentNullException"><paramref name="input"/> is null.</exception>
-        /// <exception cref="ArgumentNullException"><paramref name="project"/> is null.</exception>
+        /// <exception cref="ArgumentNullException"><paramref name="project"/> or <paramref name="utterance"/> is null.</exception>
         /// <exception cref="RequestFailedException">The service returned an error. The exception contains details of the service error.</exception>
-        public virtual async Task<Response<AnalyzeConversationTaskResult>> AnalyzeConversationAsync(ConversationItemBase input, ConversationsProject project, ConversationAnalysisOptions options = null, CancellationToken cancellationToken = default)
+        public virtual async Task<Response<AnalyzeConversationTaskResult>> AnalyzeConversationAsync(string utterance, ConversationsProject project, AnalyzeConversationOptions options = null, CancellationToken cancellationToken = default)
         {
-            Argument.AssertNotNull(input, nameof(input));
+            Argument.AssertNotNull(utterance, nameof(utterance));
```
Comment (heaths) [inline sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/ConversationAnalysisClient.cs:85]: Use `AssertNotNullOrEmpty` for strings in most cases.

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/ConversationAnalysisClient.cs [hunk 12]
```diff
@@ -74,21 +74,22 @@ protected ConversationAnalysisClient()
...
             Argument.AssertNotNull(project, nameof(project));
 
             CustomConversationTaskParameters customConversationTaskParameters = new CustomConversationTaskParameters(project.ProjectName, project.DeploymentName) { Verbose = options?.Verbose };
```
Comment (heaths) [inline sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/ConversationAnalysisClient.cs:88]: Put object initializers on their own line e.g.,

```csharp
CustomConversationTaskParameters customConversationTaskParameters = new CustomConversationTaskParameters(project.ProjectName, project.DeploymentName)
{
  Verbose = options?.Verbose,
};
```

Note also the trailing comma. This style makes it much easier to diff and maintain over time.

**Review by heaths — CHANGES_REQUESTED** [review]
Just a few more changes. We should also talk about the AnalyzeConversation method offline now that it takes more options.

**Comment by annatisch** [comment]
> Just a few more changes. We should also talk about the AnalyzeConversation method offline now that it takes more options.

@heaths - agreed, we should revisit how we're surfacing this across languages as it's changed significantly.
Though not necessary for this preview - we should have a discussion about in time to prepare for the May release.

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/samples/Sample4_AnalyzeConversationWithOptions.md (hunk 9 above)
Comment (heaths) [reply sdk/cognitivelanguage/Azure.AI.Language.Conversations/samples/Sample4_AnalyzeConversationWithOptions.md:32]: Also, I wonder if the return should just be `Response<AnalyzeConersationResult>`. The "Task" adds nothing.

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/README.md (hunk 6 above)
Comment (ZulaMostafa) [reply sdk/cognitivelanguage/Azure.AI.Language.Conversations/README.md:115]: that's actually interesting.

**Comment by heaths** [comment]
@ZulaMostafa for the spelling errors:

```
./sdk/cognitivelanguage/Azure.AI.Language.Conversations/api/Azure.AI.Language.Conversations.netstandard2.0.cs:369:71 - Unknown word (Petabit)
./sdk/cognitivelanguage/Azure.AI.Language.Conversations/api/Azure.AI.Language.Conversations.netstandard2.0.cs:456:66 - Unknown word (Picometer)
##[error]Spelling errors detected. To correct false positives or learn about spell checking see: https://aka.ms/azsdk/engsys/spellcheck
```

Open .vscode/cspell.json and change this section:

```diff
    {
      "filename": "**/sdk/cognitivelanguage/**/*",
      "words": [
        "apim",
        "nigiri",
-       "gnas"
+       "qnas",
+       "petabit",
+       "picometer" 
      ]
    },
```

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/samples/Sample4_AnalyzeConversationWithOptions.md (hunk 9 above)
Comment (ZulaMostafa) [reply sdk/cognitivelanguage/Azure.AI.Language.Conversations/samples/Sample4_AnalyzeConversationWithOptions.md]: `AnalyzeConversationTaskResult` is used as a return type in the rest client. both are different.

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/ConversationAnalysisClient.cs [hunk 13]
```diff
@@ -82,10 +82,13 @@ protected ConversationAnalysisClient()
         /// <exception cref="RequestFailedException">The service returned an error. The exception contains details of the service error.</exception>
         public virtual async Task<Response<AnalyzeConversationTaskResult>> AnalyzeConversationAsync(string utterance, ConversationsProject project, AnalyzeConversationOptions options = null, CancellationToken cancellationToken = default)
         {
-            Argument.AssertNotNull(utterance, nameof(utterance));
+            Argument.AssertNotNullOrEmpty(utterance, nameof(utterance));
```
Comment (heaths) [inline sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/ConversationAnalysisClient.cs:85]: Should also document the `ArgumentException` if `utterance` is an empty string. Same below as well.

**Review by heaths — APPROVED** [review]
Just the spelling and the exception and I can merge it.

**Review by heaths — APPROVED** [review]
